src/suiryoku/locomotion/model/robot.cpp

```diff
@@ -47,4 +50,240 @@ keisan::Angle<double> Robot::get_tilt() const
   return tilt + tilt_center;
 }
 
+void Robot::localize(bool initial_localization)
+{
+  if (num_particles == 0 || initial_localization) {
+    init_particles(initial_localization);
+  } else {
+    update_motion();
+  }
+
+  if (projected_objects.empty()) {
+    return;
+  }
+
+  calculate_weight();
+  resample_particles();
+
+  estimate_position();
+  print_estimate_position();
+}
+
+void Robot::init_particles(bool initial_localization)
+{
+  particles.clear();
+  if (initial_localization) {
+    std::cout << "INIT PARTICLES BASED ON LAST POSE" << std::endl;
+
+    num_particles = 1000;
+    std::random_device xrd, yrd;
+    std::normal_distribution<double> xrg(position.x, xvar), yrg(position.y, yvar);
+
+    for (int i = 0; i < num_particles; ++i) {
+      Particle new_particle;
+      new_particle.position = keisan::Point2(xrg(xrd), yrg(yrd));
+      new_particle.orientation = orientation;
+      new_particle.weight = 1.0 / num_particles;
+
+      particles.push_back(new_particle);
```

Comment:
Minor improvement, you can resize the particles size first instead of push_back to reduce memory reallocations.
```suggestion
    particles.resize(num_particles);
    for (int i = 0; i < num_particles; ++i) {
      Particle new_particle;
      new_particle.position = keisan::Point2(xrg(xrd), yrg(yrd));
      new_particle.orientation = orientation;
      new_particle.weight = 1.0 / num_particles;

      particles[i] = new_particle;
```

Check other loop that can be improved this way also.